presto-main/src/main/java/com/facebook/presto/sql/planner/GroupingOperationRewriter.java

```diff
@@ -0,0 +1,328 @@
+    }
+
+    @Override
+    public Expression rewriteWhenClause(WhenClause node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        return updateTypeIfRewritten(treeRewriter.defaultRewrite(node, context), node);
+    }
+
+    @Override
+    public Expression rewriteCoalesceExpression(CoalesceExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        return updateTypeIfRewritten(treeRewriter.defaultRewrite(node, context), node);
+    }
+
+    @Override
+    public Expression rewriteInListExpression(InListExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        return updateTypeIfRewritten(treeRewriter.defaultRewrite(node, context), node);
+    }
+
+    @Override
+    public Expression rewriteFunctionCall(FunctionCall node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        FunctionCall rewrittenFunctionCall = (FunctionCall) updateTypeIfRewritten(treeRewriter.defaultRewrite(node, context), node);
+
+        if (rewrittenFunctionCall != node) {
+            Signature signature = analysis.getFunctionSignature(node);
+            analysis.addFunctionSignature(rewrittenFunctionCall, signature);
+
+            return rewrittenFunctionCall;
+        }
+
+        return node;
+    }
+
+    @Override
+    public Expression rewriteLambdaExpression(LambdaExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        return updateTypeIfRewritten(treeRewriter.defaultRewrite(node, context), node);
+    }
+
+    @Override
+    public Expression rewriteLikePredicate(LikePredicate node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        return updateTypeIfRewritten(treeRewriter.defaultRewrite(node, context), node);
+    }
+
+    @Override
+    public Expression rewriteInPredicate(InPredicate node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        return updateTypeIfRewritten(treeRewriter.defaultRewrite(node, context), node);
+    }
+
+    @Override
+    public Expression rewriteArrayConstructor(ArrayConstructor node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        return updateTypeIfRewritten(treeRewriter.defaultRewrite(node, context), node);
+    }
+
+    @Override
+    public Expression rewriteSubscriptExpression(SubscriptExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        return updateTypeIfRewritten(treeRewriter.defaultRewrite(node, context), node);
+    }
+
+    @Override
+    public Expression rewriteCast(Cast node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        return updateTypeIfRewritten(treeRewriter.defaultRewrite(node, context), node);
+    }
+
+    @Override
+    public Expression rewriteTryExpression(TryExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        return updateTypeIfRewritten(treeRewriter.defaultRewrite(node, context), node);
```

Comment:
I agree, this was bad design on my part. I've refactored the code such that no mutation of the Analysis object occurs. See 15c59d02f5eb5c5bf87a8736a36bd03b3ae9323f.